espnet2/tasks/spk.py

```diff
@@ -0,0 +1,304 @@
+import argparse
+from typing import Callable, Collection, Dict, List, Optional, Tuple
+
+import numpy as np
+import torch
+from typeguard import check_argument_types, check_return_type
+
+from espnet2.asr.encoder.abs_encoder import AbsEncoder
+from espnet2.asr.frontend.abs_frontend import AbsFrontend
+from espnet2.asr.frontend.default import DefaultFrontend
+from espnet2.asr.frontend.windowing import SlidingWindow
+from espnet2.asr.specaug.abs_specaug import AbsSpecAug
+from espnet2.asr.specaug.specaug import SpecAug
+from espnet2.layers.abs_normalize import AbsNormalize
+from espnet2.layers.global_mvn import GlobalMVN
+from espnet2.layers.utterance_mvn import UtteranceMVN
+from espnet2.spk.encoder.rawnet3_encoder import RawNet3Encoder
+from espnet2.spk.espnet_model import ESPnetSpeakerModel
+from espnet2.spk.loss.aamsoftmax import AAMSoftmax
+from espnet2.spk.pooling.abs_pooling import AbsPooling
+from espnet2.spk.pooling.chn_attn_stat_pooling import ChnAttnStatPooling
+from espnet2.spk.projector.abs_projector import AbsProjector
+from espnet2.spk.projector.rawnet3_projector import RawNet3Projector
+from espnet2.tasks.abs_task import AbsTask
+from espnet2.torch_utils.initialize import initialize
+from espnet2.train.class_choices import ClassChoices
+from espnet2.train.collate_fn import CommonCollateFn
+from espnet2.train.preprocessor import (
+    AbsPreprocessor,
+    CommonPreprocessor,
+    SpkPreprocessor,
+)
+from espnet2.train.trainer import Trainer
+from espnet2.utils.get_default_kwargs import get_default_kwargs
+from espnet2.utils.nested_dict_action import NestedDictAction
+from espnet2.utils.types import int_or_none, str2bool, str_or_none
+
+# Check and understand
+frontend_choices = ClassChoices(
+    name="frontend",
+    classes=dict(
+        default=DefaultFrontend,
+        sliding_window=SlidingWindow,
+        raw=AbsFrontend,
+    ),
+    type_check=AbsFrontend,
+    default="default",
+    optional=True,
+)
+
+specaug_choices = ClassChoices(
+    name="specaug",
+    classes=dict(specaug=SpecAug),
+    type_check=AbsSpecAug,
+    default=None,
+    optional=True,
+)
+
+normalize_choices = ClassChoices(
+    name="normalize",
+    classes=dict(
+        global_mvn=GlobalMVN,
+        utterance_mvn=UtteranceMVN,
+    ),
+    type_check=AbsNormalize,
+    default=None,
+    optional=True,
+)
+
+# add more choices (e.g., ECAPA-TDNN)
+encoder_choices = ClassChoices(
+    name="encoder",
+    classes=dict(
+        # conformer=ConformerEncoder, #TODO: add.
+        rawnet3=RawNet3Encoder,
+    ),
+    type_check=AbsEncoder,
+    default="rawnet3",
+)
+
+pooling_choices = ClassChoices(
+    name="pooling",
+    classes=dict(
+        # TODO: implement additional aggregators
```

Comment:
It would be better if you add todos with your name in it so that we can trace it easier later~ (e.g., TODO(Jee-weon)